File: benchmark/sieve.rs
fn init_sieve(arr: &mut [i64], mut i: i64, size: i64) {
    while i < size {
        arr[i as usize] = 1;
        i += 1;
    }
}

fn clear_multiples(arr: &mut [i64], factor: i64, mut num: i64, size: i64) {
    while num < size {
        arr[num as usize] = 0;
        num += factor;
    }
}

fn find_next_prime(arr: &[i64], mut i: i64, limit: i64) -> i64 {
    while i <= limit {
        if arr[i as usize] == 1 {
            return i;
        }
        i += 1;
    }
    0
}

fn isqrt(n: i64, mut guess: i64) -> i64 {
    loop {
        let next = (guess + n / guess) / 2;
        if next >= guess {
            return guess;
        }
        guess = next;
    }
}

fn count_primes(arr: &[i64], size: i64) -> i64 {
    let mut count: i64 = 1;
    let mut i: i64 = 0;
    while i < size {
        if arr[i as usize] == 1 {
            count += 1;
        }
        i += 1;
    }
    count
}

fn run_sieve(limit: i64) -> i64 {
    let size = (limit + 1) / 2;
    let mut arr = vec![0i64; size as usize];
    init_sieve(&mut arr, 0, size);
    arr[0] = 0;

    let q = isqrt(limit, limit / 2);

    let mut factor: i64 = 3;
    while factor <= q {
        let next = find_next_prime(&arr, factor / 2, q / 2);
        if next == 0 {
            break;
        }
        let actual_factor = next * 2 + 1;
        let start = actual_factor * actual_factor / 2;
        clear_multiples(&mut arr, actual_factor, start, size);
        factor = actual_factor + 2;
    }

    count_primes(&arr, size)
}

fn main() {
    let count = run_sieve(1000000);
    println!("Primes found: {}", count);
}

File: benchmark/Sieve.java
public class Sieve {

    static void initSieve(long[] arr, int i, int size) {
        while (i < size) {
            arr[i] = 1;
            i++;
        }
    }

    static void clearMultiples(long[] arr, int factor, int num, int size) {
        while (num < size) {
            arr[num] = 0;
            num += factor;
        }
    }

    static int findNextPrime(long[] arr, int i, int limit) {
        while (i <= limit) {
            if (arr[i] == 1) {
                return i;
            }
            i++;
        }
        return 0;
    }

    static int isqrt(int n, int guess) {
        while (true) {
            int next = (guess + n / guess) / 2;
            if (next >= guess) {
                return guess;
            }
            guess = next;
        }
    }

    static long countPrimes(long[] arr, int size) {
        long count = 1;
        for (int i = 0; i < size; i++) {
            count += arr[i];
        }
        return count;
    }

    static long runSieve(int limit) {
        int size = (limit + 1) / 2;
        long[] arr = new long[size];
        initSieve(arr, 0, size);
        arr[0] = 0;

        int q = isqrt(limit, limit / 2);
        int factor = 3;

        while (factor <= q) {
            int next = findNextPrime(arr, factor / 2, q / 2);
            if (next == 0) {
                break;
            }
            int actualFactor = next * 2 + 1;
            int start = actualFactor * actualFactor / 2;
            clearMultiples(arr, actualFactor, start, size);
            factor = actualFactor + 2;
        }

        return countPrimes(arr, size);
    }

    public static void main(String[] args) {
        System.out.println("Primes found: " + runSieve(1000000));
    }
}
